### PR TITLE
Add Tarjan and Lengauer's (almost) linear time dominators algorithm

### PR DESCRIPTION
**Pull request by james-d-mitchell**
~~This work in progress~~

### CODE REVIEW

**Comment by james-d-mitchell** [comment]
@marinaanagno @samanthaharper @finnbuck

**Comment by james-d-mitchell** [comment]
The following things still need to be done (hoops to be jumped through):

- remove the quadratic time version of `Dominators`. This could be put into the test file as a stand alone function, for testing purposes. 
- rename `Dominators2` to `Dominators` and update the documentation etc. Also ensure that `Dominators2` is a proper GAP method (like `Dominators` currently is).
- make `DominatorTree` a proper GAP operation (like `Dominators` currently is).
- add documentation for `DominatorTree`
- add tests for `DominatorTree`

Also we might consider:

- reenabling the `lastlinked` optimisation
- making `buckets` flat again.

**Comment by james-d-mitchell** [comment]
There are still a couple of things to do in this PR, so no need for review just yet.

**Comment by marinaanagno** [comment]
Thanks @james-d-mitchell! Let me know if there's anything I could do!

**Comment by james-d-mitchell** [comment]
@marinaanagno when you have a moment, can you please review this PR? Once you've done that I'll squash the commits into one again (I left them separate for the time being so that you can see what I did today), and add you and Sam as co-authors for the commit.

File: doc/oper.xml
Comment (marinaanagno) [inline doc/oper.xml:1587]: My knowledge in complexity is almost zero right now, but since we visit all the nodes in the digraph but only the edges of the subdigraph induced by the reachable vertices, shouldn't we get that "n is the number of nodes and m is the number of edges in the digraph induced by the vertices reachable from the root vertex"?

Comment (marinaanagno) [inline doc/oper.xml:1587]: Lacking knowledge in complexity, but I was thinking that `DominatorTree` performs a dfs so it will have worst case complexity O(m + n), but `Dominators` only runs through the vertices that are reachable from the root vertex and so I cannot find a case, in which we would visit any of the edges not belonging in the subdigraph induced by the reachable nodes. Definitely not sure about this though!

**Review by marinaanagno — COMMENTED** [review]
The code seems to agree with the theory stated in the paper and seems to be giving right results. Since this is a project I've been involved though and since I am a newcomer a quick look from another pair of eyes would be great! My only comment is on the complexity, but again it is very possible that I might be missing something.

File: doc/oper.xml
Comment (marinaanagno) [inline doc/oper.xml:1587]:
```suggestion
    <M>O(mlog n)</M> where <M>m</M>  and <M>n</M> are the
```

File: doc/oper.xml
Comment (marinaanagno) [inline doc/oper.xml:1588]:
```suggestion
    number of edges and nodes in the subdigraph induced by the nodes in <A>digraph</A> respectively.
```

Comment (marinaanagno) [inline doc/oper.xml:1588]:
```suggestion
    number of edges and number of nodes in the subdigraph induced by the nodes in <A>digraph</A>
```

File: doc/oper.xml
Comment (marinaanagno) [inline doc/oper.xml:1589]:
```suggestion
    reachable from <A>root</A>, respectively.
```

**Review by marinaanagno — CHANGES_REQUESTED** [review]
Everything looks good, I only suggested a change in the phrasing in the complexity part of the documentation

**Comment by james-d-mitchell** [comment]
Thanks @marinaanagno much appreciated. I wonder if the wording about the complexity in the manual entry for `DominatorTree` should be adjusted in the same way you suggested for `Dominators`?

**Comment by marinaanagno** [comment]
@james-d-mitchell you're absolutely right! Sorry I didn't annotate that!

File: gap/oper.gi
Comment (marinaanagno) [inline gap/oper.gi:1780]: This function is recursive and I remember that we said that functions in a package shouldn't be recursive, since there is a limit on how many recursions GAP can "handle".

**Review by marinaanagno — COMMENTED** [review]
Sorry for the delay, it occurred to me after I had a look. `Compress` is a recursive function and I'm afraid we've said that we have to make it non-recursive before merging.

**Comment by james-d-mitchell** [comment]
> Sorry for the delay, it occurred to me after I had a look. `Compress` is a recursive function and I'm afraid we've said that we have to make it non-recursive before merging.

Good point! I’ll have to fix this, maybe not today though 😎

**Comment by marinaanagno** [comment]
> Sorry for the delay, it occurred to me after I had a look. `Compress` is a recursive function and I'm afraid we've said that we have to make it non-recursive before merging.

After discussing this issue with @james-d-mitchell we figured out that compress will only run log2 of the number of vertices and no digraph that can be calculated using GAP will hit the recursion limit for compress

**Comment by wilfwilson** [comment]
Something that reviewers should have in mind while reviewing: what should the return value be?

File: gap/oper.gi
Comment (wilfwilson) [inline gap/oper.gi:1842]: I noticed that the pseudocode in the paper instead has
```suggestion
      idom[w] := idom[idom[w]];
```
I just tried putting it like that and it doesn't work! So this is a bug in the paper, right?

**Review by wilfwilson — CHANGES_REQUESTED** [review]
Thanks @marinaanagno and @samanthaharper for your work on this! I've looked at it for quite a while, including reading parts of the paper (although I don't have the time to read the whole thing).

I see that you've followed the pseudocode of the paper quite closely, so I'm pretty much happy to believe that you've done that correctly (having checked a few bits myself) 🙂. The number of tests give me confidence that you have.

My one slight concern about the code is the following. At points, the paper defines things in terms of the vertices _less than_ another vertex (i.e. `5 < 7`). Seeing as you're most familiar with the code, I would just like you to convince yourselves that you are **not** relying on either the list of out-neighbours or the list of in-neighbours of a vertex being sorted. Because they aren't guaranteed to be sorted (although in-neighbours *usually* are sorted into ascending order).

Some quick testing indicates this isn't an issue, but I'd appreciate it if you could think about it.

I've 'requested changes' for my suggestions on the documentation.

As for the question 'what should the function return' which @james-d-mitchell brought up on the call today, I now understand that to mean 'should `DominatorsTree` exist as a user-facing function'? And I don't particularly mind about this either way. But if you keep it, I'm happy for it to return a record as it currently does (just please document what those things mean, mathematically 🙂 ).

File: gap/oper.gi
Comment (wilfwilson) [reply gap/oper.gi:1780]: See https://github.com/digraphs/Digraphs/pull/336#issuecomment-755413562

**Comment by marinaanagno** [comment]
> Thanks @marinaanagno and @samanthaharper for your work on this! I've looked at it for quite a while, including reading parts of the paper (although I don't have the time to read the whole thing).
> 
> I see that you've followed the pseudocode of the paper quite closely, so I'm pretty much happy to believe that you've done that correctly (having checked a few bits myself) 🙂. The number of tests give me confidence that you have.
> 
> My one slight concern about the code is the following. At points, the paper defines things in terms of the vertices _less than_ another vertex (i.e. `5 < 7`). Seeing as you're most familiar with the code, I would just like you to convince yourselves that you are **not** relying on either the list of out-neighbours or the list of in-neighbours of a vertex being sorted. Because they aren't guaranteed to be sorted (although in-neighbours _usually_ are sorted into ascending order).
> 
> Some quick testing indicates this isn't an issue, but I'd appreciate it if you could think about it.
> 
> I've 'requested changed' for my suggestions on the documentation.
> 
> As for the question 'what should the function return' which @james-d-mitchell brought up on the call today, I now understand that to mean 'should `DominatorsTree` exist as a user-facing function'? And I don't particularly mind about this either way. But if you keep it, I'm happy for it to return a record as it currently does (just please document what those things mean, mathematically 🙂 ).

Thank you so much for the review @wilfwilson, I will look carefully at all the changes you suggested probably tomorrow :)

File: gap/oper.gi
Comment (marinaanagno) [reply gap/oper.gi:1842]: Yes, this is a bug in the paper that took us ages to fix! :)

File: gap/oper.gi
Comment (marinaanagno) [reply gap/oper.gi:1858]: I remember talking about this with @james-d-mitchell at some point and saying that we need to think about it, and it is probably something we need to reconsider! Thanks for pointing that out! :)

**Comment by james-d-mitchell** [comment]
@wilfwilson @marinaanagno @samanthaharper I think this is good to go, except possibly: I'm not sure if we ever arrived at a decision about whether or not a vertex should dominate itself, any thoughts?

**Comment by marinaanagno** [comment]
@james-d-mitchell thanks for this! According to the theory a vertex is not dominated by itself and if there's no reason to include it in the list of dominators, I would suggest we make the algorithm agree with the theory. :)

**Comment by wilfwilson** [comment]
That was my feeling too.

**Comment by james-d-mitchell** [comment]
Thanks @marinaanagno and @wilfwilson I've updated the PR to not include a vertex as a dominator of itself (I think). The only other thing that I think was pending was whether or not `DominatorTree` should be user facing. I think it should because it's really the fundamental thing computed by the algorithm, and `Dominators` is just a useful additional feature. I've updated `DominatorTree` to only return the preorder and the immediate dominators, both of which are used by the current `Dominators` method.

**Review by marinaanagno — APPROVED** [review]

**Comment by marinaanagno** [comment]
@james-d-mitchell looks good to me! As far as `DominatorTree` is concerned, as a user working on this project, I almost always use `DominatorTree`, because it gives me much more information, such as preorder values, semidominators (which is now removed) etc. Unless there is any particular reason to prevent us from making it user facing, and since documentation for it already exists, I don't see any reason to not have it available for the user to use. :) Also, semidominators was a pretty important notion for this algorithm and it was very useful to have it as part of the output in `DominatorTree` and maybe keeping it in the record would be an option, however I'm not really sure whether it is commonly used as a concept, so maybe there's indeed no reason to keep it in the record. :)

**Review by wilfwilson — APPROVED** [review]
Nice! I've just updated this with the master branch; I'll merge if the CI passes again (which it should!).

**Comment by marinaanagno** [comment]
Thanks so much @wilfwilson ! :)

**Pull request merged**
